kubespawner/utils.py

```diff
@@ -27,3 +27,130 @@ def generate_hashed_slug(slug, limit=63, hash_length=6):
+
+    changes_dict = _get_k8s_model_dict(model_type, changes)
+    for key, value in changes_dict.items():
+        if key not in target.attribute_map:
+            raise ValueError("The attribute 'changes' ({}) contained '{}' not modeled by '{}'.".format(type(changes).__name__, key, model_type.__name__))
+        
+
+        # If changes are passed as a dict, they will only have a few keys/value
+        # pairs representing the specific changes. If the changes parameter is a
+        # model instance on the other hand, the changes parameter will have a
+        # lot of default values as well. These default values, which are also
+        # falsy, should not use to override the target's values.
+        if isinstance(changes, dict) or value:
+            if getattr(target, key):
+                if logger and target_name:
+                    logger.warning("Overriding KubeSpawner.{}'s value '{}' with '{}'.".format(target_name, getattr(target, key), value))
+            setattr(target, key, value)
+
+    return target
+
+def get_k8s_model(model_type, model_dict):
+    """
+    Returns an instance of type specified model_type from an model instance or
+    represantative dictionary.
+    """
+    model_dict = copy.deepcopy(model_dict)
+
+    if isinstance(model_dict, model_type):
+        return model_dict
+    elif isinstance(model_dict, dict):
+        # convert the dictionaries camelCase keys to snake_case keys
+        _map_dict_keys_to_model_attributes(model_type, model_dict)
+        # use the dictionary keys to initialize a model of given type
+        return model_type(**model_dict)
+    else:
+        raise AttributeError("Expected object of type 'dict' (or '{}') but got '{}'.".format(model_type.__name__, type(model_dict).__name__))
+
+def _get_k8s_model_dict(model_type, model):
+    """
+    Returns a dictionary representation of a provided model type
+    """
+    model = copy.deepcopy(model)
+
+    if isinstance(model, model_type):
+        return model.to_dict()
+    elif isinstance(model, dict):
+        return _map_dict_keys_to_model_attributes(model_type, model)
+    else:
+        raise AttributeError("Expected object of type '{}' (or 'dict') but got '{}'.".format(model_type.__name__, type(model).__name__))
+
+def _map_dict_keys_to_model_attributes(model_type, model_dict):
+    """
+    Maps a dict's keys to the provided models attributes using its attribute_map
+    attribute. This is (always?) the same as converting camelCase to snake_case.
+    Note that the function will not influence nested object's keys.
+    """
+
+    # it is important to iterate over a copy of the dictionaries keys, as they
+    # will be updated in the loop
+    for key in list(model_dict.keys()):
+        model_dict[_get_k8s_model_attribute(model_type, key)] = model_dict.pop(key)
```

Comment:
I think this should create a new dictionary instead of mutating one in-place

Reply:
Thanks for considering this @minrk.

I'm intending to use this helper twice as written out below, this code is from PR Part 4.

```python
    # the assignments to notebook_container and pod.spec is pointless atm, but won't be if I update the utility function update_k8s_model to be get_updated_copy_of_k8s_model or similar.
    if extra_container_config:
        notebook_container = update_k8s_model(
            target=notebook_container,
            changes=extra_container_config,
            logger=logger,
            target_name="extra_container_config",
        )

    # ...

    if extra_pod_config:
        pod.spec = update_k8s_model(
            target=pod.spec,
            changes=extra_pod_config,
            logger=logger,
            target_name="extra_pod_config",
        )
```

I reasoned that we are already making a lot of changes to `notebook_container` and `pod.spec`, and pod.spec will be a huge object relatively so I was thinking I wasn't sure that I could do a deepcopy and retain good enough performance. Also note that anything else than a deepcopy of objects like pod.spec containing a lot of nested reference objects would be misleading I think.

Reply:
I trust your experience on this, so I'll go with whatever you think is best and be right on it, the change won't be causing much work at all if we do it!

Reply:
Hmmm... `target_name` isn't right, it's really `changes_name`